thruster-core-async-await/src/middleware.rs

```diff
@@ -1,12 +1,26 @@
 use std::boxed::Box;
 use futures::future::Future;
 use std::pin::Pin;
+use crate::errors::ThrusterError;
 
+#[cfg(not(feature = "thruster_error_handling"))]
+pub type MiddlewareResult<C> = C;
+#[cfg(not(feature = "thruster_error_handling"))]
 pub type MiddlewareReturnValue<T> = Pin<Box<Future<Output=T> + Send + Sync>>;
+#[cfg(not(feature = "thruster_error_handling"))]
 pub type MiddlewareNext<C> = Box<Fn(C) -> Pin<Box<Future<Output=C> + Send + Sync>> + Send + Sync>;
-
+#[cfg(not(feature = "thruster_error_handling"))]
 type MiddlewareFn<C> = fn(C, Box<((Fn(C) -> Pin<Box<Future<Output=C> + Send + Sync>>) + 'static + Send + Sync)>) -> Pin<Box<Future<Output=C> + Send + Sync>>;
 
+#[cfg(feature = "thruster_error_handling")]
+pub type MiddlewareResult<C> = Result<C, ThrusterError<C>>;
```

Comment:
Could this use the `Context + Error` traits instead of `ThrusterError`?

Reply:
I thought this was a great idea, so I started digging through what would need to change -- it'd require a full stack change since:

`MiddlewareResult` is used in `MiddlewareFn` is used in `Middleware` is used in `Chain` is used in `Node`, so it's a pretty pervasive change. I get the value of this though, so perhaps it makes sense to just implement `from` or `into` for custom error types?

Reply:
I did the same and discovered similar. If you make the error a `Box<dyn Error>` you don't need the extra type parameter. If you add in `downcast-rs` on `trait Error` you can still stash your own errors. This still isn't as nice as being able to use your own error type, but it gets a bit closer.

Reply:
Possibly adding a cause: Option<Box<dyn std::error::Error>> to `ThrusterError` along with the `into` would be good enough?